### PR TITLE
[ONNX] Deprecated the example_outputs param from torch.onnx.export() function.

### PR DESCRIPTION
**Pull request by hwangdeyu**
`example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

### CODE REVIEW

File: torch/onnx/__init__.py [hunk 1]
```diff
@@ -197,6 +197,7 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
             Since this optimization adjusts model initializers, it will be disabled if
             export_params = False or keep_initializers_as_inputs = True.
         example_outputs (T or a tuple of T, where T is Tensor or convertible to Tensor, default None):
+            [Deprecated and ignored. Will be removed in next PyTorch release, do this part in internal erporte function],
```
Comment (jiafatom) [inline torch/onnx/__init__.py:200]: nit: "in internal erporte" -> "in the internal"

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -483,8 +483,20 @@ def _model_to_graph(model, args, verbose=False,
...
     if isinstance(model, torch.jit.ScriptModule) or isinstance(model, torch.jit.ScriptFunction):
-        assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule or " \
-                                            "ScriptFunction."
+        input_args = copy.deepcopy(args)
+        input_kwargs = {}
+        if input_args and isinstance(input_args[-1], dict):
+            input_kwargs = input_args[-1]
+            input_args = input_args[:-1]
+        try:
+            model_copy = copy.deepcopy(model)
```
Comment (jiafatom) [inline torch/onnx/utils.py:492]: why we need deepcopy model here? This seems expensive for big model. Can we just use `model` as below?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -483,8 +483,20 @@ def _model_to_graph(model, args, verbose=False,
                             module=module)
     from torch.onnx.symbolic_helper import _onnx_shape_inference
     if isinstance(model, torch.jit.ScriptModule) or isinstance(model, torch.jit.ScriptFunction):
-        assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule or " \
-                                            "ScriptFunction."
+        input_args = copy.deepcopy(args)
+        input_kwargs = {}
+        if input_args and isinstance(input_args[-1], dict):
+            input_kwargs = input_args[-1]
```
Comment (jiafatom) [inline torch/onnx/utils.py:489]: There is same logic in `_decide_input_format` to compute `input_kwargs` and `input_args`, can we use a function to reuse the code?

File: torch/onnx/utils.py (hunk 2 above)
Comment (hwangdeyu) [reply torch/onnx/utils.py:492]: My concern is that it may have the inplace ops in model. So we need deep copy this model.

File: torch/onnx/utils.py (hunk 3 above)
Comment (hwangdeyu) [reply torch/onnx/utils.py:489]: Thanks for this comment.
The `_decide_input_format` is used before calling `_model_to_graph()`.https://github.com/pytorch/pytorch/blob/66846701b85715747090d79c1c3d793731747328/torch/onnx/utils.py#L702 for decideing the format of input.
In this part, we only need get the kwargs and args so we couldn't reuse that.

File: torch/onnx/__init__.py (hunk 1 above)
Comment (hwangdeyu) [reply torch/onnx/__init__.py:200]: Thanks for catching it. Done!

File: torch/onnx/utils.py (hunk 2 above)
Comment (jiafatom) [reply torch/onnx/utils.py:491]: For inplace ops issue, we can the input value since it would change. Why we need copy the model?

File: torch/onnx/__init__.py [hunk 4]
```diff
@@ -197,6 +197,7 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
             Since this optimization adjusts model initializers, it will be disabled if
             export_params = False or keep_initializers_as_inputs = True.
         example_outputs (T or a tuple of T, where T is Tensor or convertible to Tensor, default None):
+            [Deprecated and ignored. Will be removed in next PyTorch release, do this part in the internal function],
```
Comment (neginraoof) [inline torch/onnx/__init__.py:200]: I don't think we need to mention internal function. The internal function's argument will be unused as well.

Comment (neginraoof) [inline torch/onnx/__init__.py:200]: I don't think we need to mention internal function. The internal function's argument will be unused as well.

File: test/onnx/test_utility_funs.py [hunk 5]
```diff
@@ -101,25 +99,6 @@ def forward(self, x, y, t):
         for node in graph.nodes():
             assert node.kind() != "onnx::SplitToSequence"
 
-    def test_output_list(self):
```
Comment (neginraoof) [inline test/onnx/test_utility_funs.py:104]: Why is this test removed?

**Review by neginraoof — CHANGES_REQUESTED** [review]

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -483,8 +482,20 @@ def _model_to_graph(model, args, verbose=False,
                             module=module)
     from torch.onnx.symbolic_helper import _onnx_shape_inference
     if isinstance(model, torch.jit.ScriptModule) or isinstance(model, torch.jit.ScriptFunction):
-        assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule or " \
-                                            "ScriptFunction."
+        input_args = copy.deepcopy(args)
```
[the remark this answers is not in the document]
Comment (hwangdeyu) [reply torch/onnx/utils.py:485]: Thanks, good point for code structure.

File: test/onnx/test_utility_funs.py (hunk 5 above)
Comment (hwangdeyu) [reply test/onnx/test_utility_funs.py:104]: As this test is used for handling incorrect format for `example_outputs`. If we want to remove this parameter, it might not need this any more.

File: torch/onnx/utils.py (hunk 2 above)
Comment (hwangdeyu) [reply torch/onnx/utils.py:491]: Sorry, I have added the deepcopy logical for inplace ops issue in [line 485](https://github.com/pytorch/pytorch/blob/d853f72a228c7ed61e3831549f35772854606680/torch/onnx/utils.py#L485).

The reason why we copy the model is for some cases weights or attributes changed after running the model once.

Comment (hwangdeyu) [reply torch/onnx/utils.py]: I have tried to test the weight changed case. 
Such as batchnorm training mode, if we don't use deepcopy() to copy a new model, the weight `batch_model.running_mean` and `batch_model.running_var` will be changed. However, the unit tests will not be failed, cause we export a model with the new weight with `batch_model.running_mean` and `batch_model.running_var`, but we don't expect this behaviour.

File: torch/onnx/__init__.py (hunk 4 above)
Comment (hwangdeyu) [reply torch/onnx/__init__.py]: Thanks, done!

File: torch/onnx/utils.py [hunk 7]
```diff
@@ -445,6 +447,24 @@ def _get_named_param_dict(graph, params):
...
     return _params_dict
 
+def _get_example_outputs(model, args):
+    input_args = copy.deepcopy(args)
+    input_kwargs = {}
+    if input_args and isinstance(input_args[-1], dict):
+        input_kwargs = input_args[-1]
+        input_args = input_args[:-1]
+    try:
+        model_copy = copy.deepcopy(model)
```
Comment (jiafatom) [inline torch/onnx/utils.py:457]: Can we do deepcopy only if `training is True`?

Comment (hwangdeyu) [reply torch/onnx/utils.py]: Make sense. I removed deepcopy here. Thanks!

**Review by jiafatom — COMMENTED** [review]
LGTM! For the test issue, please sync with @neginraoof . Thank you!

**Review by neginraoof — APPROVED** [review]
Thanks! LGTM.

File: torch/onnx/utils.py [hunk 8]
```diff
@@ -483,14 +496,14 @@ def _model_to_graph(model, args, verbose=False,
                             module=module)
     from torch.onnx.symbolic_helper import _onnx_shape_inference
     if isinstance(model, torch.jit.ScriptModule) or isinstance(model, torch.jit.ScriptFunction):
-        assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule or " \
-                                            "ScriptFunction."
-        if isinstance(example_outputs, list):
+        if example_outputs is None:
+            example_outputs = _get_example_outputs(model, args)
+        elif isinstance(example_outputs, list):
+            # example_outpus specified
```
Comment (BowenBao) [inline torch/onnx/utils.py:502]: @hwangdeyu This branch does not replicate the old behavior, due to missing calling the above snippet

```python
    if isinstance(example_outputs, (torch.Tensor, int, float, bool)):
        example_outputs = (example_outputs,)
```

Comment (BowenBao) [reply torch/onnx/utils.py:502]: I will try to edit the PR that submits to master, and tag you for review.

Comment (hwangdeyu) [reply torch/onnx/utils.py:502]: > I will try to edit the PR that submits to master, and tag you for review.

Sure, that's a necessary modification. Thanks!
